modules/smtp/modules.php

```diff
@@ -2291,3 +2452,102 @@ function recip_count_check($headers, $omod) {
         Hm_Msgs::add('ERRMessage contains more than the maximum number of recipients, proceed with caution');
     }
 }}
+
+/**
+ * @subpackage smtp/functions
+ */
+if (!hm_exists('send_scheduled_message')) {
+function send_scheduled_message($handler, $imap, $msg, $server_id) {
+    $msg_headers = $imap->get_message_headers($msg['uid']);
+    $imap_details = Hm_IMAP_List::dump($server_id);
+    if (empty($imap_details)) {
+        return;
+    }
+    try {
+        if (empty($msg_headers['X-Schedule'])) {
+            return;
+        }
+        if (new DateTime($msg_headers['X-Schedule']) <= new DateTime()) {
+            $profile = Hm_Profiles::get($msg_headers['X-Profile-ID']);
+            if (!$profile) {
+                $profiles = Hm_Profiles::search('server', $imap_details['server']);
+
+                if (!$profiles) {
+                    Hm_Debug::add(sprintf('ERRCannot find profiles corresponding with IMAP server: %s', $imap_details['server']));
+                    return;
+                }
+                $profile = $profiles[0];
+            }
+
+            $smtp = Hm_SMTP_List::connect($profile['smtp_id'], false);
+
+            if (smtp_authed($smtp)) {
+                if (isset($msg_headers['X-Delivery'])) {
+                    $from_params = 'RET=HDRS';
+                    $recipients_params = 'NOTIFY=SUCCESS,FAILURE';
+                } else {
+                    $from_params = '';
+                    $recipients_params = '';
+                }
+                $recipients = [];
+                foreach (['To', 'Cc', 'Bcc'] as $fld) {
+                    if (array_key_exists($fld, $msg_headers)) {
+                        $recipients = array_merge($recipients, Hm_MIME_Msg::find_addresses($msg_headers[$fld]));
+                    }
+                }
+                $msg_content = $imap->get_message_content($msg['uid'], 0);
+                $from = process_address_fld($msg_headers['From']);
+
+                $err_msg = $smtp->send_message($from[0]['email'], $recipients, $msg_content, $from_params, $recipients_params);
+
+                if (!$err_msg) {
+                    if ($imap->message_action('DELETE', [$msg['uid']])) {
+                        $imap->message_action('EXPUNGE', [$msg['uid']]);
+                    }
+                    save_sent_msg($handler, $server_id, $imap, $imap_details, $msg_content, $msg['uid'], false);
+                    return true;
+                }
+            }
+        }
+    } catch (Exception $e) {
+        Hm_Debug::add(sprintf('ERRCannot send message: %s', $msg_headers['subject']));
+    }
+    return;
+}}
+
+if (!hm_exists('reschedule_message_sending')) {
+function reschedule_message_sending($handler, $imap, $msg_id, $folder, $new_date, $server_id) {
```

Comment:
Use modules.php only for handle or output modules. Move the functions to its own functions.php file to keep the existing structure in other modules (e.g. imap and core).

Reply:
@amaninyumu1 What do you think?

Reply:
Yes, this was resolved.